bug: APIv2.2 protocols error when they try to explicitly re-use returned tips

<!--
  Thanks for taking the time to file an issue! Please make sure you've read the
  "Opening Issues" section of our Contributing Guide:

  https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-issues

  To ensure your issue can be addressed quickly, please fill out the sections
  below to the best of your ability!
-->

# Overview

#4681  fixed a bug where `pipette.return_tip(); pipette.pick_up_tip()` would erroneously re-use the tip it just returned.

However, it appears to have introduced another bug in its place: `pipette.return_tip(location); pipette.pick_up_tip(location)` fails with an "out of tips" error.

# Current behavior / steps to reproduce

The following protocol:

```python
metadata = {"apiLevel": "2.2"}

def run(protocol):
    tip_rack = protocol.load_labware("opentrons_96_tiprack_300ul", 1)
    pipette = protocol.load_instrument("p300_single_gen2", mount="left")
    well = tip_rack["A1"]
    pipette.pick_up_tip(well)
    pipette.return_tip(well)
    pipette.pick_up_tip(well) # Error
    pipette.return_tip(well)
```

Fails with: 

```
Traceback (most recent call last):
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/execute.py", line 108, in _run_python
    exec('run(context)', new_globs, new_locs)
  File "<string>", line 1, in <module>
  File "tip_tracking_test.py", line 9, in run
    pipette.pick_up_tip(well) # Error
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/util.py", line 289, in _check_version_wrapper
    return decorated_obj(*args, **kwargs)
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/instrument_context.py", line 655, in pick_up_tip
    tiprack.use_tips(target, self.channels)
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/labware.py", line 728, in use_tips
    assert all([well.has_tip for well in target_wells]),\
AssertionError: Opentrons 96 Tip Rack 300 µL on 1 is out of tips

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/maxpm/.pyenv/versions/3.8.0/bin/opentrons_simulate", line 10, in <module>
    sys.exit(main())
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/simulate.py", line 484, in main
    runlog, maybe_bundle = simulate(
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/simulate.py", line 316, in simulate
    execute.run_protocol(protocol, context)
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/execute.py", line 131, in run_protocol
    _run_python(protocol, context)
  File "/Users/maxpm/.pyenv/versions/3.8.0/lib/python3.8/site-packages/opentrons/protocol_api/execute.py", line 119, in _run_python
    raise ExceptionInProtocolError(e, tb, str(e), frame.lineno)
opentrons.protocol_api.execute.ExceptionInProtocolError: AssertionError [line 9]: Opentrons 96 Tip Rack 300 µL on 1 is out of tips
```
<!--
  Use this section to describe your bug at a high level. Please include any
  issues you can find that may be related.
-->

# Expected behavior

* If I do `pick_up_tip()`, _without_ a location, the pipette should pick up the first available "clean" tip (or column of tips). That is, it should pick up the first available tip that has not already been picked up and returned.
* If I do `pick_up_tip(location)`, _with_ a location, the pipette should do what I tell it to and pick up that tip, even if it's "dirty."

--- a/api/src/opentrons/protocol_api/labware.py
+++ b/api/src/opentrons/protocol_api/labware.py
@@ -725,8 +725,17 @@ class Labware(DeckItem):
         num_tips = min(len(target_column) - well_idx, num_channels)
         target_wells = target_column[well_idx: well_idx + num_tips]
 
-        assert all([well.has_tip for well in target_wells]),\
-            '{} is out of tips'.format(str(self))
+        # In API version 2.2, we no longer reset the tip tracker when a tip
+        # is dropped back into a tiprack well. This fixes a behavior where
+        # subsequent transfers would reuse the dirty tip. However, sometimes
+        # the user explicitly wants to use a dirty tip, and this check would
+        # raise an exception if they tried to do so.
+        # An extension of work here is to have separate tip trackers for
+        # dirty tips and non-present tips; but until then, we can avoid the
+        # exception.
+        if self._api_version < APIVersion(2, 2):
+            assert all([well.has_tip for well in target_wells]),\
+                '{} is out of tips'.format(str(self))
 
         for well in target_wells:
             well.has_tip = False

--- a/api/tests/opentrons/protocol_api/test_labware.py
+++ b/api/tests/opentrons/protocol_api/test_labware.py
@@ -5,6 +5,7 @@ import pytest
 from opentrons.protocol_api import labware, MAX_SUPPORTED_VERSION
 from opentrons.system.shared_data import load_shared_data
 from opentrons.types import Point, Location
+from opentrons.protocols.types import APIVersion
 
 test_data = {
     'circular_well_json': {
@@ -289,6 +290,18 @@ def test_select_next_tip():
     assert next_five == well_list[16]
     next_eight = tiprack.next_tip(8)
     assert next_eight == well_list[16]
+
+    # you can reuse tips infinitely on api level 2.2
+    tiprack.use_tips(well_list[0])
+    tiprack.use_tips(well_list[0])
+
+    # you can't on api level 2.1 or previous
+    early_tr = labware.Labware(labware_def,
+                               Location(Point(0, 0, 0), 'Test Slot'),
+                               api_level=APIVersion(2, 1))
+    early_tr.use_tips(well_list[0])
+    with pytest.raises(AssertionError):
+        early_tr.use_tips(well_list[0])
 
 
 def test_previous_tip():

